fsw/src/crypto.c

```diff
@@ -2647,46 +2648,252 @@ static int32 Crypto_PDU(char* ingest,TC_t* tc_frame)
+            Crypto_saPrint(&temp_SA);
+        #endif
+
+        // Verify SA SPI configurations
+        // Check if SA is operational
+        if ((temp_SA.sa_state != SA_OPERATIONAL) && (status == OS_SUCCESS))
+        {
+            OS_printf(KRED "Error: SA is not operational! \n" RESET);
+            status = OS_ERROR;
+        }
+        // Check if this VCID exists / is mapped to TCs
+        if ((temp_SA.gvcid_tc_blk[temp_tc_header.vcid].mapid != TYPE_TC) && (status = OS_SUCCESS))
+        {	
+            OS_printf(KRED "Error: SA does not map this VCID to TC! \n" RESET);
+            status = OS_ERROR;
+        }
+
+        // If a check is invalid so far, can return
+        if (status != OS_SUCCESS)
+        {
+            return status;
+        }
+
+        // Determine mode
+        // Clear
+        if ((temp_SA.est == 0) && (temp_SA.ast == 0))
+        {
+            #ifdef DEBUG
+                OS_printf(KBLU "Creating a TC - CLEAR! \n" RESET);
+            #endif
+
+            // Total length of buffer to be malloced
+            // Ingest length + segment header (1) + spi_index (2) + some variable length fields
+            // TODO
+            *enc_frame_len = in_frame_length + 1 + 2 + temp_SA.shplf_len;
+
+            #ifdef TC_DEBUG
+                OS_printf(KYEL "DEBUG - Total TC Buffer to be malloced is: %d bytes\n" RESET, *enc_frame_len);
+                OS_printf(KYEL "\tlen of TF\t = %d\n" RESET, in_frame_length);
+                OS_printf(KYEL "\tsegment hdr\t = 1\n" RESET);
+                OS_printf(KYEL "\tspi len\t\t = 2\n" RESET);
+                OS_printf(KYEL "\tsh pad len\t = %d\n" RESET, temp_SA.shplf_len);
+            #endif
+
+            // Accio buffer
+            uint8 *enc_frame = (const uint8 *)malloc(*enc_frame_len * sizeof (unsigned char));
+            CFE_PSP_MemSet(enc_frame, 0, *enc_frame_len);
+
+            // Copy original TF header
+            CFE_PSP_MemCpy(enc_frame, in_frame, TC_FRAME_PRIMARYHEADER_SIZE);
+            // Set new TF Header length
+            // Recall: Length field is one minus total length per spec
+            *(enc_frame+2) = ((*(enc_frame+2) & 0xFC) | (((*enc_frame_len - 1) & (0x0300)) >> 8));
+            *(enc_frame+3) = ((*enc_frame_len - 1) & (0x00FF));
+
+            #ifdef TC_DEBUG
+                OS_printf(KYEL "Printing updated TF Header:\n\t");
+                for (int i=0; i<TC_FRAME_PRIMARYHEADER_SIZE; i++)
+                {
+                    OS_printf("%02X", *(enc_frame+i));
+                }
+                // Recall: The buffer length is 1 greater than the field value set in the TCTF
+                OS_printf("\n\tNew length is 0x%02X\n", *enc_frame_len-1);
+                OS_printf(RESET);
+            #endif
+
+            /*
+            ** Start variable length fields
+            */
+
+            uint16_t index = TC_FRAME_PRIMARYHEADER_SIZE;
+            // Set Secondary Header flags
+            *(enc_frame + index) = 0xFF;
```

Comment:
secondary header is optional ... Please add //todo - should only add secondary header if it already exists -- it should be copied as is, not set to FF.